clients/common/host_alloc.cpp

```diff
@@ -0,0 +1,229 @@
+
+    ptrdiff_t n_bytes = -1; // unknown
+
+    FILE* fp = popen("cat /proc/meminfo", "r");
+    if(fp == NULL)
+    {
+        return n_bytes;
+    }
+
+    static const char* mem_token     = "MemFree";
+    static auto*       mem_free_type = getenv("HIPBLAS_CLIENT_ALLOC_AVAILABLE");
+    if(mem_free_type)
+    {
+        mem_token = "MemAvail"; // MemAvailable
+    }
+    int mem_token_len = strlen(mem_token);
+
+    while(fgets(buf, BUF_MAX, fp) != NULL)
+    {
+        // set env HIPBLAS_CLIENT_ALLOC_AVAILABLE to use MemAvailable if too many SKIPS occur
+        if(!strncmp(buf, mem_token, mem_token_len))
+        {
+            sscanf(buf, "%*s %td", &n_bytes); // kB assumed as 3rd column and ignored
+            n_bytes *= 1024;
+            break;
+        }
+    }
+
+    int status = pclose(fp);
+    if(status == -1)
+    {
+        return -1;
+    }
+    else
+    {
+        return n_bytes;
+    }
+
+#endif
+}
+
+inline bool host_mem_safe(size_t n_bytes)
+{
+#if defined(HIPBLAS_BENCH)
+    return true; // roll out to hipblas-bench when CI does perf testing
+#else
+    static auto* no_alloc_check = getenv("HIPBLAS_CLIENT_NO_ALLOC_CHECK");
+    if(no_alloc_check)
+    {
+        return true;
+    }
+
+    constexpr size_t threshold = 100 * 1024 * 1024; // 100 MB
+
+    static size_t client_ram_limit = 0;
+
+    static int once = [&] {
+        auto* alloc_limit = getenv("HIPBLAS_CLIENT_RAM_GB_LIMIT");
+        if(alloc_limit)
+        {
+            size_t mem_limit;
+            client_ram_limit = sscanf(alloc_limit, "%zu", &mem_limit) == 1 ? mem_limit : 0;
+            client_ram_limit <<= 30; // B to GB
+        }
+        return 0;
+    }();
+
+    if(n_bytes > threshold)
+    {
+        if(client_ram_limit)
+        {
+            if(host_bytes_allocated() + n_bytes > client_ram_limit)
+            {
+                std::cout << "Warning: skipped allocating " << n_bytes << " bytes ("
+                          << (n_bytes >> 30) << " GB) as total would be more than client limit ("
+                          << (client_ram_limit >> 30) << " GB)" << std::endl;
+
+                return false;
+            }
+        }
+
+        ptrdiff_t avail_bytes = host_bytes_available(); // negative if unknown
+        if(avail_bytes >= 0 && n_bytes > avail_bytes)
+        {
+            std::cout << "Warning: skipped allocating " << n_bytes << " bytes (" << (n_bytes >> 30)
+                      << " GB) as more than free memory (" << (avail_bytes >> 30) << " GB)"
+                      << std::endl;
+
+            // we don't try if it looks to push load into swap
+            return false;
+        }
+    }
+    return true;
+#endif
+}
+
+void* host_malloc(size_t size)
+{
+    if(host_mem_safe(size))
+    {
+        void* ptr = malloc(size);
+
+        static int value = -1;
+
+        static auto once = false;
+        if(!once)
+        {
+            auto* alloc_byte_str = getenv("HIPBLAS_CLIENT_ALLOC_FILL_HEX_BYTE");
```

Comment:
This is just used for our own debugging, I suppose? @TorreZuk this question might be aimed at you.